Use webpack asset modules

jnoordsij opened the pull request:
Replace deprecated `file-loader` with Webpack asset modules, as described in https://webpack.js.org/guides/asset-modules/.

Fixes #2914 and fixes #3249 for my setup.

thecrypticace commented:
Because people are fairly reliant on `file-loader` we'll want a way to _opt-in_ to asset modules. I suggest adding a config option for this (so you can enable it via `mix.options`). You'll be able to pull from the config there using `mix.config`.

Maybe something like `mix.options({ assetModules: true })` though I'm very open to other suggestions for naming.

thecrypticace commented:
I do want to _require_ them in the next major version. It should've been caught when we upgraded Mix to v6 but sadly I missed it. 🙃

jnoordsij commented:
Thanks for your feedback! I've added a config option now that toggles the new behavior. Also I adjusted tests to use this new toggle by default and copied the two main tests which test these rules to test both old and new version. Finally force-pushed to remove the commit that deleted the `file-loader` dependency, as it is still used.

Let me know if you have any more suggestions and/or improvements!

thecrypticace commented:
Used this as an opportunity to do a tiny refactor of the rules code. I've always felt it was far too messy. I _think_ it's better but it doesn't feel 100% where I want it to be. But better is better in any case. Thanks again for doing this! ❤️ 

I think there's a few other things I'll want to take care of before tagging a patch release but I'm aiming to cut a release on either Friday or Saturday.

jnoordsij reviewed (commented):
Great, thanks for merging. Looking forward to the release! Noticed two possibly incorrect references, might wanna check/fix those first.

jnoordsij commented:
> Used this as an opportunity to do a tiny refactor of the rules code. I've always felt it was far too messy. I _think_ it's better but it doesn't feel 100% where I want it to be. But better is better in any case. Thanks again for doing this! ❤️
> 
> I think there's a few other things I'll want to take care of before tagging a patch release but I'm aiming to cut a release on either Friday or Saturday.

Any update on the next patch release? 😇

thecrypticace commented:
Heh yeah sorry! Life got in the way. Soon™ — I want to see if I can fix the `[object Module]` in Vue 2 with these first.

jnoordsij commented:
> Heh yeah sorry! Life got in the way. Soon™ — I want to see if I can fix the `[object Module]` in Vue 2 with these first.

Great! Would be nice to have before June 1st, when Node 17 reaches EOL, so I can upgrade my builds to Node 18 by then.

thecrypticace commented:
Ah yesss it does fix it! 💯 

Just more evidence that I really shouldn't have missed this the first time around when I upgraded Mix to webpack 5. lol
